cmd/client-keystone-auth/main.go

```diff
@@ -70,59 +70,92 @@ func promptForString(field string, r io.Reader, show bool) (result string, err e
+	if err != nil && err.Error() == "unexpected newline" {
+		return "", nil
+	}
+
 	return result, err
 }
 
-// prompt pulls keystone auth url, domain, username and password from stdin,
-// if they are not specified initially (i.e. equal "").
-func prompt(url string, domain string, user string, project string, password string) (gophercloud.AuthOptions, error) {
+// prompt pulls missing auth parameters from stdin, if they were not specified initially (i.e. equal "")
+func prompt(options *gophercloud.AuthOptions) error {
 	var err error
-	var options gophercloud.AuthOptions
 
-	if url == "" {
-		url, err = promptForString("Keystone Auth URL", os.Stdin, true)
+	if options.IdentityEndpoint == "" {
+		options.IdentityEndpoint, err = promptForString("Keystone auth URL", os.Stdin, true)
 		if err != nil {
-			return options, err
+			return err
 		}
-	}
 
-	if domain == "" {
-		domain, err = promptForString("domain name", os.Stdin, true)
-		if err != nil {
-			return options, err
+		if options.IdentityEndpoint == "" {
+			return fmt.Errorf("Keystone url must be specified")
 		}
 	}
 
-	if user == "" {
-		user, err = promptForString("user name", os.Stdin, true)
+	if options.DomainID == "" && options.DomainName == "" {
+		options.DomainName, err = promptForString("domain name (leave empty to set domain id instead)", os.Stdin, true)
 		if err != nil {
-			return options, err
+			return err
+		}
+
+		if options.DomainName == "" {
+			options.DomainID, err = promptForString("domain ID", os.Stdin, true)
+			if err != nil {
+				return err
+			}
+		}
+
+		if options.DomainID == "" && options.DomainName == "" {
+			return fmt.Errorf("either domain ID or domain name must be specified")
 		}
 	}
 
-	if project == "" {
-		project, err = promptForString("project name", os.Stdin, true)
+	if options.UserID == "" && options.Username == "" {
+		options.Username, err = promptForString("user name (leave empty to set user id instead)", os.Stdin, true)
 		if err != nil {
-			return options, err
+			return err
+		}
+
+		if options.Username == "" {
+			options.UserID, err = promptForString("user ID", os.Stdin, true)
+			if err != nil {
+				return err
+			}
+		}
+
+		if options.UserID == "" && options.Username == "" {
+			return fmt.Errorf("either user ID or user name must be specified")
 		}
 	}
 
-	if password == "" {
-		password, err = promptForString("password", nil, false)
+	if options.TenantID == "" && options.TenantName == "" {
+		options.TenantName, err = promptForString("project name (leave empty to set project id instead)", os.Stdin, true)
 		if err != nil {
-			return options, err
+			return err
+		}
+
+		if options.TenantName == "" {
+			options.UserID, err = promptForString("project ID", os.Stdin, true)
+			if err != nil {
+				return err
+			}
+		}
+
+		if options.TenantID == "" && options.TenantName == "" {
+			fmt.Fprintf(os.Stderr, "Neither project ID nor project name have been specified. Proceed with domain-scoped token.")
```

Comment:
Is there any reason we might want to allow domain name do be set to 'default' if there is no input?